core/src/main/java/org/infinispan/reactive/publisher/impl/LocalPublisherManagerImpl.java

```diff
@@ -169,70 +284,116 @@ public void segmentsLost(IntSet lostSegments) {
-                     innerFlowable = innerFlowable.filter(i -> !keysToExclude.contains(toKeyFunction.apply(i)));
-                  }
+         resultFlowable = exactlyOnceParallel(set, keysToExclude, toKeyFunction, segments, transformer,
+               listener, concurrentSegments);
+      } else {
+         resultFlowable = exactlyOnceSequential(set, keysToExclude, toKeyFunction, segments, transformer,
+               listener, concurrentSegments);
+      }
+      return exactlyOnceHandleLostSegments(finalizer.apply(resultFlowable), listener);
+   }
 
-                  CompletionStage<R> stage = transformer.apply(innerFlowable);
-                  CompletableFuture<R> future = stage.toCompletableFuture();
-                  // This will always be true unless there is a store
-                  if (future.isDone()) {
-                     if (listener.segmentsLost.contains(segment)) {
-                        return CompletableFutures.<R>completedNull();
-                     }
-                     return future;
-                  }
+   protected <R> CompletionStage<PublisherResult<R>> exactlyOnceHandleLostSegments(CompletionStage<R> finalValue, SegmentListener listener) {
+      return handleLostSegments(finalValue, listener);
+   }
 
-                  return stage.thenCompose(value -> {
-                     // This means the segment was lost in the middle of processing
-                     if (listener.segmentsLost.contains(segment)) {
-                        return CompletableFutures.completedNull();
-                     }
-                     return CompletableFuture.completedFuture(value);
-                  });
-               }).sequential();
+   /**
+    * This method iteratively submits a task to operate on the cpu bound thread pool for each segment that is available.
+    * The tasks perform a type of work stealing where they attempt to retrieve the next available segment and process
+    * them as fast as possible. It is entirely possible that a given task is never submitted due to the other tasks
+    * completing all the segments asynchronously. After the main thread has submitted all the tasks it will attempt
+    * to steal a segment and run it if possible and if it can will subsequently attempt to complete all remaining
+    * segments in the same fashion as the other threads. NOTE that this behavior is not normally how reactive
+    * streams are done as given operations are not normally performed until the returned Flowable is subscribed to, but
+    * for performance reasons this method eagerly publishes entries.
+    * <p>
+    * The results of each segment data will then be published each as a single result in the returned Flowable. Due
+    * to the results being retrieved eagerly it is entirely possible that if the Subscriber of the Flowable is slow
+    * that that results queue up. But due to that the fact that results are reduced to single values for each segment
```

Comment:
In some scenario they could be bigger (we wouldn't know). The biggest ones would be something like a `collect` or `toArray`, but those methods need all the results as one anyways. However I miswrote, it is a single value for each submitted task (ie. numCPU or numCPU + 1). Not per segment, which should reduce it significantly more as well.

Reply:
Actually sorry it is per segment, got stuck thinking the wrong thing because of the previous comment - I need to look more closely at context next time :)